backupccl: handle pre-23.1 tenant metadata in backup manifests

stevendanna opened the pull request:
In 23.1 the canonical field for the ID and DataState of a tenant moved, with the old fields being renamed to DeprecatedID and DeprecatedDataState.

While we updated the code that _writes_ the manifest. We did not update all of the code that _reads_ the manifest. As a result, code reading the manifest during RESTORE or SHOW BACKUP would attempt to restore tenants with an ID of 0.

Here, we copy over the relevant tenant metadata from the deprecated field to the new field in GetTenants(), remove stray direct accesses to Tenants, remove use of MustMakeTenantID from SHOW, and add test coverage.

Epic: none

Release note (bug fix): Fixes a bug in which older backups of tenants could not be restored. Only Cockroach Labs has such backups.

Release justification: High priority bug fix to ensure tenants can be restored from older backups.

stevendanna committed:
backupccl: handle pre-23.1 tenant metadata in backup manifests

In 23.1 the canonical field for the ID and DataState of a tenant
moved, with the old fields being renamed to DeprecatedTenant and
DeprecatedID.

While we updated the code that _writes_ the manifest. We did not
update all of the code that _reads_ the manifest. As a result, code
reading the manifest during RESTORE or SHOW BACKUP would attempt to
restore tenants with an ID of 0.

Here, we copy over the relevant tenant metadata from the deprecated
field to the new field in GetTenants(), remove stray direct accesses
to Tenants, remove use of MustMakeTenantID from SHOW, and add test coverage.

NB: An alternative here would be to do this upgrade in our
readManifest routine in the backupinfo package. I think this would
probably be better overall and may do a follow up commit to make it
so.

Epic: none

Release note (bug fix): Fixes a bug in which older backups of tenants
could not be restored. Only Cockroach Labs has such backups.

Release justification: High priority bug fix to ensure tenants can be
restored from older backups.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104245)
<!-- Reviewable:end -->

stevendanna committed:
backupccl: move tenant metadata upgrade to readManifest

Rather than upgrading the tenant metadata  on demand in GetTenants() we now
update it in the backupinfo package when reading the manifest. This
removes the need for GetTenants().

Along the way, I removed more uses of MustMakeTenantID from the backup
code as we almost never want to panic in this code.

Epic: none

Release note: None

msbutler commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -1409,6 +1422,9 @@ func getTenantInfo(
 		)
 	}
 	for i := range tenants {
+		// NB: We use MustMakeTenantID here since the data is
```
nit: what do you mean by "the data is coming from the database"?

msbutler commented on pkg/ccl/backupccl/backuppb/backup.proto:
```diff
@@ -93,6 +93,12 @@ message BackupManifest {
   repeated File files = 4 [(gogoproto.nullable) = false];
   repeated sql.sqlbase.Descriptor descriptors = 5 [(gogoproto.nullable) = false];
+  // Tenants holds metadata about any tenants contained in the
+  // backup. Callers who constructed a BackupManifest using without
```
nit: remove first "using"

msbutler commented on pkg/ccl/backupccl/backuppb/backup_test.go:
```diff
@@ -0,0 +1,167 @@
+	})
+	t.Run("copies ProtoInfo fields to SQLInfo from deprecated tenants", func(t *testing.T) {
+		manifest := backuppb.BackupManifest{
+			TenantsDeprecated: []mtinfopb.ProtoInfo{{
```
nit: i think this input case is covered above

msbutler commented on pkg/ccl/backupccl/testdata/restore_old_versions/cluster-with-tenants/v21.1.7/2023/06/02-093257.93/BACKUP-CHECKPOINT-870409758129553409-CHECKSUM:
```diff
@@ -0,0 +1 @@
+u2�
```
per slack convo, i think it's fine to drop the 21.1 test case

msbutler reviewed (approved):
LGTM! Left only nits

msbutler commented on pkg/ccl/backupccl/restore_old_versions_test.go:
```diff
@@ -445,3 +452,30 @@ func fullClusterRestoreSystemExternalConnectionsWithoutIDs(exportDir string) fun
+		err := os.Symlink(exportDir, filepath.Join(tmpDir, "foo"))
+		require.NoError(t, err)
+		sqlDB.CheckQueryResults(t, fmt.Sprintf("SELECT count(*) FROM [SHOW BACKUP LATEST IN '%s'] WHERE object_type = 'TENANT'", localFoo), [][]string{
+			{"2"},
```
nit: perhaps it's worth checking that the right tenant names are being shown, and not just the count

stevendanna replied:
Because these were created on 22.2, they don't have names in the manifest, only IDs.  We could check the IDs, but in some sense I don't care what they are provided they don't OOM and get restored.

stevendanna committed:
backupccl: remove 21.1-only test case

We don't support backups from this version so have decided against
keeping this test in place.

Epic: none

Release note: None

stevendanna commented:
bors r=msbutler
